physical/raft/raft.go

```diff
@@ -652,23 +624,151 @@ func (b *RaftBackend) DisableUpgradeMigration() (bool, bool) {
 	return b.autopilotConfig.DisableUpgradeMigration, true
 }
 
+// RunRaftWalVerifier runs a raft log store verifier in the background, if configured to do so.
+// This periodically writes out special raft logs to verify that the log store is not corrupting data.
+func (b *RaftBackend) RunRaftWalVerifier(ctx context.Context) {
+	if !b.verifierEnabled() {
+		return
+	}
+
+	go func() {
+		ticker := time.NewTicker(b.verificationInterval())
+		defer ticker.Stop()
+
+		logger := b.logger.Named("raft-wal-verifier")
+
+		for {
+			select {
+			case <-ticker.C:
+				err := b.applyVerifierCheckpoint()
+				if err != nil {
+					logger.Error("error applying verification checkpoint", "error", err)
+				}
+				logger.Debug("sent verification checkpoint")
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (b *RaftBackend) applyVerifierCheckpoint() error {
+	data := make([]byte, 1)
+	data[0] = byte(verifierCheckpointOp)
+
+	b.permitPool.Acquire()
+	b.l.RLock()
+
+	var err error
+	applyFuture := b.raft.Apply(data, 0)
+	if e := applyFuture.Error(); e != nil {
+		err = e
+	}
+
+	// TODO: do we need to wait for the response?
+	_ = applyFuture.Response()
+	b.l.RUnlock()
+	b.permitPool.Release()
+
+	return err
+}
+
+// isLogVerifyCheckpoint is the verifier.IsCheckpointFn that can decode our raft logs for
+// their type.
+func isLogVerifyCheckpoint(l *raft.Log) (bool, error) {
+	if len(l.Data) < 1 {
+		// Shouldn't be possible! But no need to make it an error if it wasn't one
+		// before.
+		return false, nil
+	}
+
+	if l.Data[0] == byte(verifierCheckpointOp) {
+		return true, nil
+	}
+
+	return false, nil
```

Comment:
I have a feeling you copied this from my code in Consul but...

We should check that `len(l.Data) == 1` rather than `>= 1` since that is always true and it reduces some false positives.

We should also sniff the extensions field for the magic value because it's possible that this could return a false positive above for all the same reasons in the comment further up.

Reply:
Indeed I did. Since this and `isVerifierLog()` from `logVerificationChunkingShim` are doing the same thing, I factored out that logic into a common method and called it in both places.